File: ios/Classes/FlutterNeoShieldPlugin.swift
import Flutter
import UIKit

/// FlutterNeoShieldPlugin — iOS platform implementation.
///
/// Provides native memory allocation, secure wipe operations,
/// RASP checks, and screen protection.
public class FlutterNeoShieldPlugin: NSObject, FlutterPlugin {
    private var secureStorage: [String: Data] = [:]

    // Screen Shield
    private let screenProtector = ScreenProtector()
    private let screenshotDetector = ScreenshotDetector()
    private let recordingDetector = ScreenRecordingDetector()
    private let appSwitcherGuard = AppSwitcherGuard()
    private var screenEventSink: FlutterEventSink?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let memoryChannel = FlutterMethodChannel(
            name: "com.neelakandan.flutter_neo_shield/memory",
            binaryMessenger: registrar.messenger()
        )
        let raspChannel = FlutterMethodChannel(
            name: "com.neelakandan.flutter_neo_shield/rasp",
            binaryMessenger: registrar.messenger()
        )
        let screenChannel = FlutterMethodChannel(
            name: "com.neelakandan.flutter_neo_shield/screen",
            binaryMessenger: registrar.messenger()
        )
        let screenEventChannel = FlutterEventChannel(
            name: "com.neelakandan.flutter_neo_shield/screen_events",
            binaryMessenger: registrar.messenger()
        )

        let instance = FlutterNeoShieldPlugin()
        registrar.addMethodCallDelegate(instance, channel: memoryChannel)
        registrar.addMethodCallDelegate(instance, channel: raspChannel)
        registrar.addMethodCallDelegate(instance, channel: screenChannel)
        screenEventChannel.setStreamHandler(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        // Allow methods that don't require arguments
        let args = call.arguments as? [String: Any]

        switch call.method {
        // Memory Shield
        case "allocateSecure":
            guard let args = args,
                  let id = args["id"] as? String,
                  let data = args["data"] as? FlutterStandardTypedData else {
                result(FlutterError(code: "INVALID_ARGS", message: "id and data required", details: nil))
                return
            }
            secureStorage[id] = Data(data.data)
            result(nil)

        case "readSecure":
            guard let args = args,
                  let id = args["id"] as? String,
                  let data = secureStorage[id] else {
                result(FlutterError(code: "NOT_FOUND", message: "No secure data found", details: nil))
                return
            }
            result(FlutterStandardTypedData(bytes: data))

        case "wipeSecure":
            let id = args?["id"] as? String
            if let id = id, var data = secureStorage[id] {
                data.resetBytes(in: 0..<data.count)
                secureStorage.removeValue(forKey: id)
            }
            result(nil)

        case "wipeAll":
            wipeAll()
            result(nil)

        // RASP Shield
        case "checkDebugger":
            result(DebuggerDetector.check())

        case "checkRoot":
            result(JailbreakDetector.check())

        case "checkEmulator":
            result(EmulatorDetector.check())

        case "checkHooks":
            result(HookDetector.check())

        case "checkFrida":
            result(FridaDetector.check())

        case "checkIntegrity":
            result(IntegrityDetector.check())

        // Screen Shield
        case "enableScreenProtection":
            let window = UIApplication.shared.windows.first { $0.isKeyWindow }
                ?? UIApplication.shared.windows.first
            result(screenProtector.enable(in: window))

        case "disableScreenProtection":
            result(screenProtector.disable())

        case "isScreenProtectionActive":
            result(screenProtector.isActive)

        case "enableAppSwitcherGuard":
            let window = UIApplication.shared.windows.first { $0.isKeyWindow }
                ?? UIApplication.shared.windows.first
            appSwitcherGuard.enable(in: window)
            result(true)

        case "disableAppSwitcherGuard":
            appSwitcherGuard.disable()
            result(true)

        case "isScreenBeingRecorded":
            result(recordingDetector.isRecording)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func wipeAll() {
        for key in secureStorage.keys {
            if var data = secureStorage[key] {
                data.resetBytes(in: 0..<data.count)
            }
        }
        secureStorage.removeAll()
    }

    /// Set up screenshot and recording detection, sending events to the Dart side.
    private func startDetection() {
        screenshotDetector.startDetecting { [weak self] in
            self?.screenEventSink?(["type": "screenshot"])
        }
        recordingDetector.startDetecting { [weak self] isCaptured in
            self?.screenEventSink?(["type": "recording", "isRecording": isCaptured])
        }
    }

    private func stopDetection() {
        screenshotDetector.stopDetecting()
        recordingDetector.stopDetecting()
    }
}

// MARK: - FlutterStreamHandler for Screen Events
extension FlutterNeoShieldPlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        screenEventSink = events
        startDetection()
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        screenEventSink = nil
        stopDetection()
        return nil
    }
}

File: android/src/main/kotlin/com/neelakandan/flutter_neo_shield/FlutterNeoShieldPlugin.kt
package com.neelakandan.flutter_neo_shield

import android.app.Activity
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/**
 * FlutterNeoShieldPlugin — Android platform implementation.
 *
 * Provides native memory allocation, secure wipe operations,
 * RASP checks, and screen protection.
 */
class FlutterNeoShieldPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    private lateinit var channel: MethodChannel
    private lateinit var raspChannel: MethodChannel
    private lateinit var screenChannel: MethodChannel
    private var screenEventChannel: EventChannel? = null
    private val secureStorage = HashMap<String, ByteArray>()
    private val debuggerDetector = com.neelakandan.flutter_neo_shield.rasp.DebuggerDetector()
    private var applicationContext: android.content.Context? = null
    private var activity: Activity? = null
    private val screenProtector = com.neelakandan.flutter_neo_shield.screen.ScreenProtector()
    private val recordingDetector = com.neelakandan.flutter_neo_shield.screen.ScreenRecordingDetector()
    private var appSwitcherGuardEnabled = false

    override fun onAttachedToEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = binding.applicationContext

        channel = MethodChannel(binding.binaryMessenger, "com.neelakandan.flutter_neo_shield/memory")
        channel.setMethodCallHandler(this)

        raspChannel = MethodChannel(binding.binaryMessenger, "com.neelakandan.flutter_neo_shield/rasp")
        raspChannel.setMethodCallHandler(this)

        screenChannel = MethodChannel(binding.binaryMessenger, "com.neelakandan.flutter_neo_shield/screen")
        screenChannel.setMethodCallHandler(this)

        screenEventChannel = EventChannel(binding.binaryMessenger, "com.neelakandan.flutter_neo_shield/screen_events")
        screenEventChannel?.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                // Android does not have native screenshot/recording callbacks
                // below API 34. Events are sent on-demand or via polling.
            }

            override fun onCancel(arguments: Any?) {}
        })
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            // Memory Shield
            "allocateSecure" -> {
                val id = call.argument<String>("id")
                val data = call.argument<ByteArray>("data")
                if (id != null && data != null) {
                    secureStorage[id] = data.copyOf()
                    result.success(null)
                } else {
                    result.error("INVALID_ARGS", "id and data are required", null)
                }
            }
            "readSecure" -> {
                val id = call.argument<String>("id")
                if (id != null && secureStorage.containsKey(id)) {
                    result.success(secureStorage[id])
                } else {
                    result.error("NOT_FOUND", "No secure data with id: $id", null)
                }
            }
            "wipeSecure" -> {
                val id = call.argument<String>("id")
                if (id != null && secureStorage.containsKey(id)) {
                    val data = secureStorage[id]!!
                    data.fill(0)
                    secureStorage.remove(id)
                    result.success(null)
                } else {
                    result.success(null)
                }
            }
            "wipeAll" -> {
                for (entry in secureStorage.values) {
                    entry.fill(0)
                }
                secureStorage.clear()
                result.success(null)
            }

            // RASP Shield
            "checkDebugger" -> {
                result.success(debuggerDetector.check())
            }
            "checkRoot" -> {
                result.success(com.neelakandan.flutter_neo_shield.rasp.RootDetector().check())
            }
            "checkEmulator" -> {
                result.success(com.neelakandan.flutter_neo_shield.rasp.EmulatorDetector().check())
            }
            "checkHooks" -> {
                val context = applicationContext
                if (context != null) {
                    result.success(com.neelakandan.flutter_neo_shield.rasp.HookDetector().check(context))
                } else {
                    // Fail closed: report as detected when context unavailable.
                    result.success(true)
                }
            }
            "checkFrida" -> {
                result.success(com.neelakandan.flutter_neo_shield.rasp.FridaDetector().check())
            }
            "checkIntegrity" -> {
                val context = applicationContext
                if (context != null) {
                    result.success(com.neelakandan.flutter_neo_shield.rasp.IntegrityDetector().check(context))
                } else {
                    // Fail closed: report as detected when context unavailable.
                    result.success(true)
                }
            }

            // Screen Shield
            "enableScreenProtection" -> {
                result.success(screenProtector.enable(activity))
            }
            "disableScreenProtection" -> {
                result.success(screenProtector.disable(activity))
            }
            "isScreenProtectionActive" -> {
                result.success(screenProtector.isActive(activity))
            }
            "enableAppSwitcherGuard" -> {
                // On Android, FLAG_SECURE already blanks the app switcher thumbnail.
                // Enabling screen protection implicitly guards the app switcher.
                val success = screenProtector.enable(activity)
                appSwitcherGuardEnabled = success
                result.success(success)
            }
            "disableAppSwitcherGuard" -> {
                // Only disable FLAG_SECURE if screen protection itself isn't active
                appSwitcherGuardEnabled = false
                // Don't clear FLAG_SECURE here — it may be set for screen protection.
                // The app switcher guard is a logical flag on Android.
                result.success(true)
            }
            "isScreenBeingRecorded" -> {
                result.success(recordingDetector.isRecordingOrMirroring(activity))
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    // ActivityAware implementation
    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        // Wipe all on detach for safety.
        for (entry in secureStorage.values) {
            entry.fill(0)
        }
        secureStorage.clear()
        channel.setMethodCallHandler(null)
        raspChannel.setMethodCallHandler(null)
        screenChannel.setMethodCallHandler(null)
        screenEventChannel?.setStreamHandler(null)
    }
}
